pkg/buildqueue/buildqueue.go

```diff
@@ -0,0 +1,568 @@
+// is guaranteed to only be written to exactly once and is then immediately
+// closed. It is up to the caller of Enqueue() to decide whether to read on the
+// channel or not.
+type BuildNotification chan bool
+
+func newBuildNotification() BuildNotification {
+	return make(BuildNotification, 1)
+}
+
+// build is what is returned by Enqueue() to provide information about the
+// build
+type build struct {
+	builder Builder
+
+	// uuid is the UUID of the Buildable
+	uuid string
+
+	// notificationChannels is a list of a channels that should be notified
+	// about the completion of the build
+	notificationChannels []BuildNotification
+
+	// internalNotification is used to notify
+	internalNotification BuildNotification
+}
+
+// createAndAddNotificationChannel creates and adds a build notification
+// channel. This function may only be called while the build is not yet in
+// progress.
+func (b *build) createAndAddNotificationChannel() BuildNotification {
+	notify := newBuildNotification()
+	b.notificationChannels = append(b.notificationChannels, notify)
+	return notify
+}
+
+// reportStatus reports the status of a build to all notification channels.
+// This function may only be called if the build is the currentBuild. It may
+// never be called while the build is still assigned to nextBuild.
+func (b *build) reportStatus(q *BuildQueue, success bool) {
+	b.internalNotification <- success
+	close(b.internalNotification)
+
+	for _, notify := range b.notificationChannels {
+		notify <- success
+		close(notify)
+	}
+}
+
+// cancel cancels a scheduled and not yet running build. q.mutex must be held.
+func (b *build) cancelScheduled(q *BuildQueue) int {
+	numBuilds := b.numFolded()
+
+	metrics.BuildQueueEntries.With(map[string]string{
+		metrics.LabelBuildQueueName:  q.name,
+		metrics.LabelBuildQueueState: metrics.BuildQueueWaiting,
+	}).Sub(float64(numBuilds))
+
+	b.reportStatus(q, false)
+
+	return numBuilds
+}
+
+// numFolded returns the number of folded builds
+func (b *build) numFolded() int {
+	return len(b.notificationChannels)
+}
+
+func newBuild(b Builder, uuid string) *build {
+	return &build{
+		builder:              b,
+		uuid:                 uuid,
+		notificationChannels: []BuildNotification{},
+		internalNotification: make(BuildNotification, 1),
+	}
+}
+
+// buildStatus is the current status of a build
+type buildStatus struct {
+	builder Builder
+
+	// curentBuild is the building information of any running build
+	currentBuild *build
+
+	// nextBuild is the building information of the next build for this
+	// UUID
+	nextBuild *build
+}
+
+func (q *BuildQueue) enqueueBuild(b Builder) (BuildNotification, bool) {
+	build, enqueued := q.serializeBuild(b)
+	notify := build.createAndAddNotificationChannel()
+
+	metrics.BuildQueueEntries.With(map[string]string{
+		metrics.LabelBuildQueueName:  q.name,
+		metrics.LabelBuildQueueState: metrics.BuildQueueWaiting,
+	}).Inc()
+
+	b.BuildQueued()
```

Comment:
Since the `BuildsDequeued()` and `Build()` callbacks are always made on the first queued `Builder` instance, for consistency I would also make the `BuildQueued()` callbacks on the same instance:
```
build.BuildQueued()
```
This would simplify managing the state of a set of builders for users in case they queue multiple instances of `Builder` with the same UUID.